expire sessions when the password is changed

warner opened the pull request:
This series of patches causes client sessions to cease being valid when the user changes their password. Any outstanding sessions must re-authenticate.

Some notes to reviewers:
- There are several cleanup/refactoring patches before the actual functional change. These were needed to make the real change easy to read.
- The term "superSessionToken" derives from the idea that each client session (expressed as a cookie, with a session duration and authentication level) is actually part of a "super session", containing multiple normal sessions. The super session lasts from one password reset to the next (or until an explicit "log me out everywhere" action, which looks very similar). Normal sessions are not allowed to extend beyond the end of their super session.
- This adds a DB fetch at the start of each wsapi call, which is regrettably unavoidable if this feature is to work correctly. It also adds a DB fetch in any path which updates the session cookie. In the long run (i.e. after 1.0), I think the DB code should be changed to make exactly one fetch at the start of the wsapi operation (retrieving the whole user row, even if some pieces are unused), and then maybe do one write at the end of the operation. This would at least avoid doing multiple reads for a single function.
- stalled-mysql-test.js was modified to match the new reality (all wsapi operations require a DB, so some previous successes are now expected failures). I had to modify wsapi_client.js to pass the 503 from /wsapi/session_context back to the caller, instead of throwing a (sync) error that killed the test. I'd appreciate careful study here, to make sure I didn't screw anything up.
- I also changed test/lib/wsapi.js to make it possible to pass in an alternate context, so the test for this feature could simulate multiple parallel sessions (changing the password in one should cause the other to stop working).
- If a session stops working because the password was changed elsewhere, we want the client to go back to the login screen and ask for the new password. My test merely checks that the expired session gets a 403, and I don't know if that's enough to test the desired behavior, or if clients might get an ugly error and _not_ be redirected back to the login screen. (since this is a wsapi call, and not a normal page fetch, I'm optimistic).

warner committed:
wsapi cleanup: factor out checkCSRF()

warner committed:
wsapi cleanup: flatten control flow a little bit

warner committed:
wsapi.authenticateSession: async-ify, combine args into a single 'options'

This includes an error-handling argument to the callback.

warner committed:
authenticate_user: factor out updateHash code

After the user is authenticated, we might update the hashed password if
the bcrypt work factor has been changed. To simplify the code in
preparation for later changes, this patch extracts the updateHash() code
out to a separate function.

Note that this function must run *after* the `res.json()` call for two
reasons. The first is to avoid slowing down the client (send answer
first, do work later). The second is to get the right session cookie
into the POST that we send to ourselves (to /wsapi/update_password) that
does the actual hash updated. The session is updated in
wsapi.authenticateSession, but the cookie isn't regenerated until the
call to `res.json()`, and the POST is sneakily grabbing the cookie out
of the response's Set-Cookie header to copy into the request-to-self.

warner committed:
enhance test/lib/wsapi.js to accept alternate contexts

This makes it possible to test two distinct sessions at the same time,
needed to exercise expiring one session when the password is changed in
a second session.

warner committed:
add checkExpiredSession check, set session.superSessionToken, add DB methods

With this change, existing session cookies are invalidated when the user
changes their password. This will also support a "log me out of
everywhere" button, which merely needs to update the superSessionToken
as if the password was changed.

This adds .superSessionToken to all sessions in
wsapi.authenticateSession, and checks it against the database row in
during a new checkExpiredSession() call (run for all /wsapi requests,
both GETs and POSTs).

All database methods which create an account or modify the password were
changed to also set .superSessionToken (to the current time).
updatePassword() added an 'invalidateSessions' argument so passwords can
be rehashed (when the bcrypt work factor changes) *without* invalidating
sessions. Finally, the database added a new method named
superSessionToken() to retrieve the current value, for comparison in
checkExpiredSession().

wsapi/update_password.js is changed to add a call to authenticateSession
after changing the password, to update the cookie with the new
.superSessionToken. This ensures that the user's new session works on
the next call after update_password.

Database rows that don't have a superSessionToken (created before the
code was upgraded to include this change) will not enforce
session-checking, so user sessions will not be spontaneously expired
when the server is upgraded past this revision.

warner committed:
update stalled-mysql tests to match new behavior

All wsapi operations now require the database (to update+check the
superSessionToken), so some tests that previously expected operations to
succeed without a database now expect them to fail (generally 503).

wsapi_client.js was changed to pass HTTP errors during
/wsapi/session_context back to the caller, so their response code can be
checked, rather than throwing an error (and preventing any other
assertions from being made).

warner committed:
test that all sessions are expired when the password is changed

benadida commented:
I think the term supersessiontoken is going to be confusing to anyone looking at the code. Let's make it more functional, e.g. password_changed_at.

benadida commented:
I'm okay with the extra DB fetch though I agree that we should aggressively work towards one query per call only. (And we already have one query per call, I think.)

benadida commented:
@lloyd should take a look at the change in stalled-mysql-tests for sure.

warner commented:
yeah, I was using "passwordChangeTime" before switching to "superSessionToken". Lloyd told me that skinny will be looking to add a "log me out everywhere" button, which would be implemented by just bumping this token, but not actually changing the password. So I used a more neutral name to avoid winding up with something that sounded weird when seen in the context of that code.

I'm happy to change it back, though.. I agree that a "super session" is not an immediately obvious concept.

benadida commented:
ah interesting, I see your reasoning now.

How about.... lastLoginReset, or something like that?

warner committed:
rename superSessionToken to lastPasswordReset, easier to understand

lloyd commented on lib/wsapi.js:
```diff
@@ -177,60 +234,44 @@ exports.setup = function(options, app) {
+    return cookieParser(req, resp, function() {
+      bodyParser(req, resp, function() {
+        cookieSessionMiddleware(req, resp, function() {
+          checkExpiredSession(req, resp, function() {
```
hmm.   don't we _only_ need to check whether the session is expired if a user is logged in?  not all of our apis require authentication.  one such case is the `authenticate` call.

warner replied:
sure.. is there a good way to tell when it's needed? I figured a hardcoded list of operation names wouldn't be a good idea..

lloyd replied:
yeah, I think we can just say if `req.session.userid` is truey (an integer), then there must exist in the sesssion a lastPasswordReset token that matches the current one in the database.  if `req.session.userid` is falsey, the session is not yet authenticated and we can skip the db hit.

lloyd replied:
in fact, I'm rather curious what is actually going on right now for un-authenticated api calls.  req.session.userid would be undefined and we'd be passing that into `db.lastPasswordReset()`

warner replied:
it's probably returning `undefined`, which would get "stored" in the session (but of course assigning `undefined` to a property is a nop). My intention with that code was to skip the comparison if the DB returns 0 or `undefined`, since that would mean the DB row pre-dates this code (and I didn't want to gratuitously expire sessions when this code first deploys). Somehow the check for that disappeared while I was merging/rebasing this.. I'll add it back.

warner replied:
hrm, bypassing the check for unauthenticated cookies changes the behavior of stalled-db situations, causing one of the stalled-mysql-test.js tests to fail.. looking into that now

warner replied:
ok, updated the test to match, should be green now

lloyd commented on lib/db/mysql.js:
```diff
@@ -65,7 +64,8 @@ if (conf.get('env') === 'test_mysql' && process.env['STALL_MYSQL_WHEN_PRESENT'])
     "id BIGINT AUTO_INCREMENT PRIMARY KEY," +
-    "passwd CHAR(64)" +
+    "passwd CHAR(64)," +
+    "lastPasswordReset BIGINT" +
```
any reason that we would not use the `TIMESTAMP` field of mysql?  like we do for staged.ts?  It shaves 4 bytes and is symmetric.  we don't get MS resolution, but I don't think we care, right?

warner replied:
true, we don't really care, although unit tests are likely to fail if the timestamp resolution can't distinguish between two calls in rapid succession, so milliseconds are easier. (if we get full second resolution, the test will probably need an artificial 2s stall to avoid false negatives).

lloyd commented on lib/db/mysql.js:
```diff
@@ -9,14 +9,13 @@
+ *    +--- user -----------------+      |*int    id       |
+ *    |*int    id                |<-----|*int    user     |
+ *    | string passwd            |      |*string address  |
+ *    | string lastPasswordReset |      | enum   type     |
```
this diagram is out of date, it's not a string anymore.

lloyd commented on lib/wsapi/authenticate_user.js:
```diff
@@ -64,59 +64,69 @@ exports.process = function(req, res) {
+            // if the work factor has changed, update the hash here.  issue #204
+            // NOTE: this runs asynchronously and will not delay the response
+            if (config.get('bcrypt_work_factor') != bcrypt.get_rounds(hash))
+              updateHash(req, res, uid, hash);
```
very nice restructuring.  that indenting was out of hand.

lloyd commented on lib/wsapi/update_password.js:
```diff
@@ -55,13 +55,24 @@ exports.process = function(req, res) {
-        db.updatePassword(req.session.userid, hash, function(err) {
+        var passwordChanged = (req.params.oldpass != req.params.newpass);
+        db.updatePassword(req.session.userid, hash, passwordChanged,
+                          function(err) {
```
hmm.  because this is a user initiated password update, is there any good reason to not reset their active sessions if they're resetting their password to the same thing?  I feel like going the other way would be safer - always resetting sessions on password update.

warner replied:
the check is actually there to accomodate the "bcrypt work factor has changed so we're going to re-hash your password for you in the background" case, which delegates to update_password (with old==new). The unit test for that case expected the sesion to remain valid, plus it seemed a bit gratuitous to expire the user's session immediately after they log in (which is when the work-factor-has-changed check occurs) when they aren't actually changing their password.

lloyd replied:
ok, that makes sense.

lloyd commented on tests/password-update-test.js:
```diff
@@ -149,6 +173,12 @@ suite.addBatch({
+    topic: wsapi.post('/wsapi/prolong_session', {}, context2),
+    "fails as expected": function(err, r) {
+      assert.strictEqual(r.code, 403);
+    }
```
beautiful tests.

warner committed:
mysql.js: update picture to match new type of lastPasswordReset

warner committed:
wsapi.js: don't check lastPasswordReset for not-yet-authenticated sessions

Specifically this should reduce the work needed by the
'authenticate_user' call by one DB read.

warner committed:
wsapi.js: if the DB doesn't have a token, don't enforce equality

This honors the preceding comment about not gratuitously expiring
innocent sessions. Somehow this clause got lost as I was
merging/rebasing this function.

warner committed:
fix stalled-mysql-test to handle not checking DB for unauthed sessions

warner committed:
mysql schema: use TIMESTAMP for lastPasswordReset, not BIGINT

Since MySQL TIMESTAMP is quantized to whole seconds, also change tests
to add a 2s stall before changing the password, to make sure
lastPasswordReset gets a new value.

warner committed:
db/mysql.js: update schema picture in the docs

jrgm commented:
verified train-2012.08.17
